File: app/src/main/java/ru/iandreyshev/parserrss/models/web/HttpRequestHandler.java
package ru.iandreyshev.parserrss.models.web;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.concurrent.TimeUnit;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;

public class HttpRequestHandler implements IHttpRequestResult {
    private static final int MAX_CONTENT_BYTES = 5242880; // 5MB

    private static final int GOOD_RESPONSE_CODE = 200;
    private static final String DEFAULT_PROTOCOL = "http://";

    private static final int READ_TIMEOUT_SEC = 2;
    private static final int CONNECTION_TIMEOUT_SEC = 2;
    private static final int WRITE_TIMEOUT_SEC = 2;

    private final OkHttpClient mClient = new OkHttpClient.Builder()
            .readTimeout(READ_TIMEOUT_SEC, TimeUnit.SECONDS)
            .connectTimeout(CONNECTION_TIMEOUT_SEC, TimeUnit.SECONDS)
            .writeTimeout(WRITE_TIMEOUT_SEC, TimeUnit.SECONDS)
            .build();
    private State mState = State.NotSend;
    private byte[] mBody;
    private Url mUrl;
    private long mMaxContentBytes = MAX_CONTENT_BYTES;

    public HttpRequestHandler(final String urlString) {
        setUrl(urlString);

        if (mState == State.BadUrl) {
            setUrl(DEFAULT_PROTOCOL + urlString);
        }
    }

    public State sendGet() {
        if (mState != State.BadUrl) {
            send(new Request.Builder()
                    .url(mUrl.getInstance())
                    .build());
        }

        return mState;
    }

    public void setMaxContentBytes(long maxBytes) {
        mMaxContentBytes = maxBytes;
    }

    @NonNull
    @Override
    public State getState() {
        return mState;
    }

    @Nullable
    @Override
    public String getResponseBodyAsString() {
        return mBody == null ? null : new String(mBody);
    }

    @Nullable
    @Override
    public byte[] getResponseBody() {
        return mBody;
    }

    @Override
    public String getUrlStr() throws NullPointerException {
        return mUrl.toString();
    }

    private void setUrl(final String url) {
        mUrl = Url.parse(url == null ? "" : url);
        mState = mUrl == null ? State.BadUrl : State.NotSend;
    }

    private void send(Request request) {
        try (final Response response = mClient.newCall(request).execute()) {
            try (final ResponseBody body = response.body()) {
                if (response.code() != GOOD_RESPONSE_CODE || body == null || body.contentLength() > mMaxContentBytes) {
                    mState = State.BadConnection;

                    return;
                }
                mBody = body.bytes();
                mState = State.Success;

            } catch (Exception ex) {
                mState = State.BadConnection;
            }

        } catch (SecurityException ex) {
            mState = State.PermissionDenied;
        } catch (Exception ex) {
            mState = State.BadConnection;
        }
    }
}

File: app/src/main/java/ru/iandreyshev/parserrss/models/web/HttpRequestHandler.kt
package ru.iandreyshev.parserrss.models.web

import okhttp3.HttpUrl
import java.util.concurrent.TimeUnit

import okhttp3.OkHttpClient
import okhttp3.Request

class HttpRequestHandler(override val urlString: String) : IHttpRequestResult {
    companion object {
        private const val MAX_CONTENT_BYTES = 5242880L // 5MB

        private const val GOOD_RESPONSE_CODE = 200
        private const val DEFAULT_PROTOCOL = "http://"

        private const val READ_TIMEOUT_SEC = 3L
        private const val CONNECTION_TIMEOUT_SEC = 3L
        private const val WRITE_TIMEOUT_SEC = 3L
    }

    enum class State {
        NotSend,
        Success,
        BadUrl,
        BadConnection,
        PermissionDenied
    }

    var maxContentBytes: Long = MAX_CONTENT_BYTES
    var readTimeoutSec: Long = READ_TIMEOUT_SEC
    var connectionTimeoutSec: Long = CONNECTION_TIMEOUT_SEC
    var writeTimeoutSec: Long = WRITE_TIMEOUT_SEC
    override var state: State = State.NotSend
        private set
    override var body: ByteArray? = null
        private set
    override val bodyAsString: String?
        get() {
            return String(body ?: return null)
        }

    private val _url: HttpUrl? = HttpUrl.parse(urlString)
            ?: HttpUrl.parse(DEFAULT_PROTOCOL + urlString)

    init {
        state = when (_url) {
            null -> State.BadUrl
            else -> State.NotSend
        }
    }

    fun sendGet(): State {
        _url ?: return State.BadUrl

        send(Request.Builder()
                .url(_url)
                .build())

        return state
    }

    private fun send(request: Request) {
        try {
            val client = OkHttpClient.Builder()
                    .readTimeout(readTimeoutSec, TimeUnit.SECONDS)
                    .connectTimeout(connectionTimeoutSec, TimeUnit.SECONDS)
                    .writeTimeout(writeTimeoutSec, TimeUnit.SECONDS)
                    .build()

            client.newCall(request).execute().use { response ->
                response.body().use { body ->
                    state = when {
                        (response.code() != GOOD_RESPONSE_CODE || body == null || body.contentLength() > maxContentBytes) ->
                            State.BadConnection
                        else -> {
                            val source = body.source()
                            source.request(maxContentBytes)
                            this.body = source.buffer().snapshot().toByteArray()
                            State.Success
                        }
                    }
                }
            }
        } catch (ex: SecurityException) {
            state = State.PermissionDenied
        } catch (ex: Exception) {
            state = State.BadConnection
        }
    }
}
